common/src/chainstate/utxo.rs

```diff
@@ -0,0 +1,462 @@
+            is_fresh,
+        }
+    }
+
+    pub(in crate::chainstate) fn new_spent(is_fresh: bool, is_dirty: bool) -> Self {
+        Self {
+            utxo: None,
+            is_dirty,
+            is_fresh,
+        }
+    }
+
+    fn utxo_mut(&mut self) -> Option<&mut Utxo> {
+        self.utxo.as_mut()
+    }
+
+    pub fn is_dirty(&self) -> bool {
+        self.is_dirty
+    }
+
+    pub fn is_fresh(&self) -> bool {
+        self.is_fresh
+    }
+
+    pub fn is_spent(&self) -> bool {
+        self.utxo.is_none()
+    }
+
+    pub fn utxo(&self) -> Option<Utxo> {
+        self.utxo.clone()
+    }
+}
+
+pub trait UtxosView {
+    /// Retrieves utxo.
+    fn get_utxo(&self, outpoint: &OutPoint) -> Option<Utxo>;
+
+    /// Checks whether outpoint is unspent.
+    fn have_utxo(&self, outpoint: &OutPoint) -> bool;
+
+    /// Retrieves the block hash of the best block in this view
+    fn get_best_block_hash(&self) -> Option<H256>;
+
+    /// Estimated size of the whole view (0 if not implemented)
+    fn estimated_size(&self) -> usize;
+
+    /// Performs bulk modification
+    fn batch_write(
+        &mut self,
+        utxos: HashMap<OutPointKey, UtxoEntry>,
+        block_hash: H256,
+    ) -> Result<(), Error>;
+
+    fn derive_cache(&self) -> UtxosCache;
+}
+
+pub fn flush_to_base<T: UtxosView>(
+    cache: UtxosCache,
+    block_hash: H256,
+    base: &mut T,
+) -> Result<(), Error> {
+    base.batch_write(cache.utxos, block_hash)
+}
+
+#[derive(Clone, Default)]
+pub struct UtxosCache<'a> {
+    parent: Option<&'a dyn UtxosView>,
+    current_block_hash: Option<H256>,
+    pub(in crate::chainstate) utxos: HashMap<OutPointKey, UtxoEntry>,
+    memory_usage: usize,
+}
+
+impl<'a> UtxosCache<'a> {
+    /// returns a copy of the UtxoEntry, given the outpoint.
+    fn get_utxo_entry(&self, outpoint: &OutPoint) -> Option<UtxoEntry> {
+        let key = OutPointKey::from(outpoint);
+
+        if let Some(res) = self.utxos.get(&key) {
+            return Some(res.clone());
+        }
+
+        // if utxo is not found in this view, use parent's `get_utxo`.
+        self.parent.map(|parent| {
+            match parent.get_utxo(outpoint) {
+                None => {
+                    // The parent only has an empty entry for this outpoint; we can consider our
+                    // version as fresh.
+                    UtxoEntry::new_spent(true, false)
+                }
+                Some(utxo) => UtxoEntry::new(utxo, false, false),
+            }
+        })
+    }
+
+    pub fn new(parent: &'a dyn UtxosView) -> Self {
+        UtxosCache {
+            parent: Some(parent),
+            current_block_hash: None,
+            utxos: HashMap::new(),
+            memory_usage: 0,
+        }
+    }
+
+    pub fn set_best_block(&mut self, block_hash: H256) {
+        self.current_block_hash = Some(block_hash);
+    }
+
+    pub fn add_utxos(
+        &mut self,
+        tx: &Transaction,
+        height: BlockHeight,
```

Comment:
There's also a case for adding UTXOs for the mempool. Do we also use height == 0 for that? If yes, then please make the block height either an Option or even more preferably use an enum that specifies a block height OR just indicates that it's in the mempool.

Reply:
added en enum:
```
pub enum UtxoType {
    /// At which height this containing tx was included in the active block chain
    BlockChain(BlockHeight),
    MemPool,
}
```

The Utxo struct will now look like this:
```
/// The Unspent Transaction Output
#[derive(Debug, Clone, Eq, PartialEq, Encode, Decode)]
pub struct Utxo {
    output: TxOutput,
    is_block_reward: bool,
    /// identifies whether the utxo is for the blockchain or for mempool.
    tx_type: UtxoType,
}
```